runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -9317,6 +9317,391 @@ static TR::Register* inlineStringHashCode(TR::Node* node, bool isCompressed, TR:
+
+   cg->stopUsingRegister(tmp);
+   cg->stopUsingRegister(tmpXMM);
+   cg->stopUsingRegister(multiplierXMM);
+
+   for (int32_t i = 0; i < 4; i++) cg->stopUsingRegister(multiplier31PowN[i]);
+   for (int32_t i = 0; i < 4; i++) cg->stopUsingRegister(hashXMM[i]);
+
+   return result;
+   }
+
+// Implement vectorizedHashCode computation with SIMD. This implementation supports multiple processor microarchitectures,
+// supporting vectorization with 128, 256, and 512 bit vectors. It also supports signed and unsigned integer element types
+// of 8-bit, 16-bit and 32-bits in size. Elements are processed iteratively as 32-bit integers and smaller element types
+// are sign/zero extended into 32-bit integers.
+//
+// This helper generates the following code to implement vectorized hashCode
+//
+//   1. Generate the main vectorized loop unrolled by a factor of 4, at highest available vector length
+//   2. Generate a second vectorized loop, not unrolled, using 128-bit vectors
+//   3. Process residue sequentially
+//     a. for (; index < length; index++) { hash = 31 * hash + arr[index]; }
+//
+// This code relies on two helpers to generate the vectorized loop and to reduce the vectorized result into the scalar
+// hashcode value.
+//
+//   1. vectorizedHashCodeLoopHelper(...)      -> Generates the vectorized loop code for given type and vector size
+//   2. vectorizedHashCodeReductionHelper(...) -> Generates code to reduce x vectors by adding all elements together
+//
+// Future works should investigate the appropriate amount of unrolling based on block hotness, array length, and
+// cache implications of unrolling. This inline intrinsic could pose significant pressure on the code-cache given
+// the large expected size of the generated code with unrolling. This may be especially apparent if the vectorized
+// hashCode algorithm is inlined multiple times in the same method.
+//
+TR::Register *
+J9::X86::TreeEvaluator::vectorizedHashCodeHelper(TR::Node *node, TR::DataType dt, TR::Node *nodeHash, bool isSigned,
+                                                 TR::CodeGenerator *cg)
+   {
+   int32_t shift = dt - TR::Int8; /* i8 -> 0, i16 -> 1, i32 -> 2 */
+
+   TR_ASSERT_FATAL(node->getChild(1)->getOpCodeValue() == TR::iconst && node->getChild(1)->getInt() == 0, "vector hashcode offset can only be const zero.");
+   TR_ASSERT_FATAL(shift >= 0 && shift <= 2, "Unsupported datatype for vectorized hashcode");
+
+   TR::Compilation *comp = cg->comp();
+   TR::VectorLength vl = TR::VectorLength128;
+
+   if (comp->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512F))
+      vl = TR::VectorLength512;
+   else if (comp->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX2))
+      vl = TR::VectorLength256;
+
+   TR::Register *address = cg->evaluate(node->getChild(0));
+   TR::Register *length = cg->evaluate(node->getChild(2));
+   TR::Register *initHash = cg->intClobberEvaluate(nodeHash);
```

Comment:
The intent I believe was to make this code compatible with String HashCode which does not have an initial hash node because it uses the value 0. I have made some changes to check if nodeHash is provided and allocate a new register for initHash if it is not provided.